### PR TITLE
Place holder for model-based LUTs

### PR DESCRIPTION
**Pull request by vbrancat**
This PR introduces a place holder for the model-based LUTs we are going to be using in the geocoded burst workflow. 
At the moment, the s1-reader provides the computation of bistatic delay and total doppler.

Specifically, this PR introduces:
1. `utils/lut.py` function to compute a total slant range and azimuth LUT as a sum of individual LUTs
2. Expose spacings of LUTs to geocoded and radar runconfig/schemas
3. Introduce a place holder for slant range and azimuth LUTs in `s1_geocode_slc.py` and `s1_resample.py`. The overall LUTs should be subsequently fed to the corresponding core modules (`geocodeSlc` and `resample`) in the workflows

### CODE REVIEW

**Comment by vbrancat** [comment]
@LiangJYu do you have time to have a look at it and check whether this is in line with the `geocodeSlc` implementation with LUTs?

**Review by seongsujeong — COMMENTED** [review]
Looks good overall. I've added few minor suggestions.

File: src/compass/defaults/s1_cslc_radar.yaml
Comment (LiangJYu) [inline src/compass/defaults/s1_cslc_radar.yaml:53]: Add same documenting comments found the geo defaults.

File: src/compass/schemas/s1_cslc_geo.yaml
Comment (LiangJYu) [inline src/compass/schemas/s1_cslc_geo.yaml:100]:
```suggestion
  # LUT spacing in x/slant range direction in meters
  x_spacing: number(min=1, required=False)
  # LUT spacing in y/azimuth direction in seconds
  y_spacing: number(min=1, required=False)
```
In line with @seongsujeong's recommendation to use scientific notation. Also add units to documenting comments.

 Also applicable to radar schema.

File: src/compass/utils/lut.py
Comment (LiangJYu) [inline src/compass/utils/lut.py:13]:
```suggestion
        LUT spacing (in pixels) along x/slant range in meters
    ystep: int
        LUT spacing (in pixels) along x/azimuth in seconds
```
Added units to be consistent with implementation in `isce3::geocode::geocodeSlc`.

File: src/compass/utils/lut.py
Comment (LiangJYu) [inline src/compass/utils/lut.py:1]:
```suggestion
def compute_geocoding_correction_lut(burst, xstep=500, ystep=50):
```
Verbosity for clarity.

File: src/compass/s1_geocode_slc.py
Comment (LiangJYu) [inline src/compass/s1_geocode_slc.py:67]:
```suggestion
    rg_correction_lut, az_correction_lut = compute_lut(burst,
```
Verbosity for clarity.

File: src/compass/s1_resample.py
Comment (LiangJYu) [inline src/compass/s1_resample.py:14]: From my understanding, I'm not sure resample is the right place for the LUT2d correction to occur in the radar workflow. Based on the `isce3::geocode::geocodeSlc` [implementation](https://github-fn.jpl.nasa.gov/isce-3/isce/pull/1089). I think the corresponding location for these corrections should eventually be in [`s1_geocode`](https://github.com/opera-adt/COMPASS/pull/48/files). I'm assuming the radar workflow will eventually be geocoded?

Another issue with putting the corrections in resample, is how will these corrections be applied to the reference datasets, since only secondary datasets get resampled?

File: src/compass/defaults/s1_cslc_geo.yaml
Comment (vbrancat) [reply src/compass/defaults/s1_cslc_geo.yaml:60]: This comment has been addressed in the topo layers PR.

Comment (yunjunz) [reply src/compass/defaults/s1_cslc_geo.yaml:60]: Scientific notation is the old way, isn't?

The old way is more readable to me.

File: src/compass/schemas/s1_cslc_geo.yaml
Comment (vbrancat) [reply src/compass/schemas/s1_cslc_geo.yaml:100]: Just making sure before applying this change. At the moment, the only corrections we have in the `s1-reader` can be computed by setting the spacing parameters in terms of number of pixels (see https://github.com/opera-adt/s1-reader/blob/71e2c9e589f9e010b4952a574fcda83806561339/src/s1reader/s1_burst_slc.py#L401). For the schema/runconfig, do we want to use meters and second? I guess inside `lut.py`, we will need to fetch the number of pixels by dividing the spacing allocated in the runconfig by the spacing of the burst in slant range and azimuth directions. We can get the range_pixel_spacing from the Sentinel1BurstSlc object but not the azimuth spacing (which either should be hard-coded or extracted at center scene?). What do you recommend @LiangJYu @seongsujeong ? I would prefer to keep it in terms of number of pixels to avoid unnecessary conversions (especially getting the spacing in the azimuth direction is not trivial)

File: src/compass/s1_resample.py
Comment (vbrancat) [reply src/compass/s1_resample.py:14]: Good point. One solution would be to remove the corrections from the radar workflow at the moment. You are correct, at a certain point we would want to geocode the burst on a geographical grid. However, for doing that, we will likely be using the `isce3.geocode.geocode_CFloat32()`. The core module does not support the use of correction LUTs at the moment. One solution would be to use `isce3.geocode.geocodeSlc()` and turn off the azimuth carrier resampling. Anyhow, I will remove the LUT computation from the radar workflow until we make a decision

**Comment by vbrancat** [comment]
@LiangJYu do you perhaps have time to take another look at this?

File: src/compass/defaults/s1_cslc_radar.yaml
Comment (LiangJYu) [inline src/compass/defaults/s1_cslc_radar.yaml:53]:
```suggestion
               # LUT spacing in x/slant range direction in meters
               x_spacing: 500
               # LUT spacing in y/azimuth direction in seconds
               y_spacing: 50
```
Adding documentation.

File: src/compass/schemas/s1_cslc_radar.yaml
Comment (LiangJYu) [inline src/compass/schemas/s1_cslc_radar.yaml:79]:
```suggestion
  # LUT spacing in x/slant range direction in meters
  x_spacing: int(min=1, required=False)
  # LUT spacing in y/azimuth direction in seconds
```
Adding units

**Review by LiangJYu — COMMENTED** [review]
Couple more comments. There are a couple comments I have not marked as "resolved" that need addressing. They're all documenting comment related.

**Comment by vbrancat** [comment]
@LiangJYu @seongsujeong could you guys have a look at this PR earliest at your convenience?

**Review by LiangJYu — APPROVED** [review]
LGTM

File: src/compass/defaults/s1_cslc_geo.yaml
Comment (seongsujeong) [inline src/compass/defaults/s1_cslc_geo.yaml:65]: Not sure if 50 **seconds** of spacing in azimuth direction makes sense. Maybe spacing in meters?

File: src/compass/defaults/s1_cslc_radar.yaml
Comment (seongsujeong) [inline src/compass/defaults/s1_cslc_radar.yaml:55]: Same as above.

File: src/compass/utils/lut.py
Comment (seongsujeong) [inline src/compass/utils/lut.py:1]: I do not think it needs to be addressed in this PR (because it would need a quite of work), but we might need to have an option to provide azimuth / range starting times by adding the arguments for these. That was necessary especially when I was defining the correction grids based on ETAD product.
If the starting time(s) is not provided, then we can bring them from `burst`

**Review by seongsujeong — COMMENTED** [review]
@vbrancat I think this PR is almost ready to be merged. I've left few comments below.

File: src/compass/schemas/s1_cslc_geo.yaml
Comment (seongsujeong) [reply src/compass/schemas/s1_cslc_geo.yaml]: I think we can use whichever makes more sense to calculate the correction, as long as we clearly mention about what that is. In case of FM mismatch rate, I stick with what ETAD ATBD is using for correction grid. (i.e. long time - short time).

Comment (vbrancat) [reply src/compass/schemas/s1_cslc_geo.yaml]: That is interesting. So ETAD has seconds both in range and azimuth directions? I guess, and @LiangJYu can correct me if I am wrong, for the GSLC core module, we have used seconds for the LUT in azimuth and meters for the one along slant range?

File: src/compass/defaults/s1_cslc_geo.yaml
Comment (vbrancat) [reply src/compass/defaults/s1_cslc_geo.yaml:65]: Your comment made me notice an inconsistency. So far, the only two corrections that we have (e.g. bistatic delay and geometrical + steering Doppler), they do require the spacings along azimuth and slant range expressed in terms of number of pixels. See https://github.com/opera-adt/s1-reader/blob/4000bc7a31d79e740212ab31495ca28ad0e2d426/src/s1reader/s1_burst_slc.py#L424 . To me, it makes sense to have it expressed in meters/seconds. How do we solve this? Shall we solve this in the s1-reader and change the steps parameters to support meters and seconds or we do the conversion internally? What do you think @LiangJYu ?

File: src/compass/utils/lut.py
Comment (vbrancat) [reply src/compass/utils/lut.py:1]: Do we need to add this in the runconfig/schema? Would it be fine if the azimuth FM rate mismatch PR would take care of this?

File: src/compass/schemas/s1_cslc_geo.yaml
Comment (seongsujeong) [reply src/compass/schemas/s1_cslc_geo.yaml]: > So ETAD has seconds both in range and azimuth directions?

Yes.

>I guess, and @LiangJYu can correct me if I am wrong, for the GSLC core module, we have used seconds for the LUT in azimuth and meters for the one along slant range?


That's my understanding about ISCE3. We use meters when defining range grid. See [here](https://github.com/opera-adt/s1-reader/blob/4000bc7a31d79e740212ab31495ca28ad0e2d426/src/s1reader/s1_burst_slc.py#L201-L209) and [here](https://github.com/opera-adt/s1-reader/blob/4000bc7a31d79e740212ab31495ca28ad0e2d426/src/s1reader/s1_reader.py#L569)
It would be great if @LiangJYu can confirm this.

Comment (LiangJYu) [reply src/compass/schemas/s1_cslc_geo.yaml]: > > So ETAD has seconds both in range and azimuth directions?
> 
> Yes.
> 
> > I guess, and @LiangJYu can correct me if I am wrong, for the GSLC core module, we have used seconds for the LUT in azimuth and meters for the one along slant range?
> 
> That's my understanding about ISCE3. We use meters when defining range grid. See [here](https://github.com/opera-adt/s1-reader/blob/4000bc7a31d79e740212ab31495ca28ad0e2d426/src/s1reader/s1_burst_slc.py#L201-L209) and [here](https://github.com/opera-adt/s1-reader/blob/4000bc7a31d79e740212ab31495ca28ad0e2d426/src/s1reader/s1_reader.py#L569) It would be great if @LiangJYu can confirm this.

Confirming ISCE3 core geocodeSlc has az in seconds and slant range in meters.

File: src/compass/defaults/s1_cslc_geo.yaml
Comment (LiangJYu) [reply src/compass/defaults/s1_cslc_geo.yaml:65]: Will these correction LUTs only be used by the core ISCE3 modules? If so then I think we should alter `burst.bistatic_delay` and `burst.geometrical_and_steering_doppler` to return LUT2d that support meters and seconds.

If there are other processes that expect these LUT2ds in pixels, we leave s1_reader untouched and change `compass.utils.lut.compute_geocoding_correction_luts` to support meters and seconds.

What do you all think?

File: src/compass/utils/lut.py
Comment (seongsujeong) [reply src/compass/utils/lut.py:1]: > Do we need to add this in the runconfig/schema? 

Depends on how we define the correction grid on COMPASS. If we are going to define the correction grid **without** the starting timing for COMPASS production, then we wouldn't need to add them in runconfig.

Maybe adding the starting time would be more relevant to `s1-reader` than to `COMPASS`. I think adding the flexible starting times would be more relevant features when we are comparing our calculation with another source of computation (e.g. ETAD)

I will definitely include more freedom in timing when adding FM rate mismatch on `s1-reader`, but not sure if I am going to touch other corrections (e.g. `bistatic_delay`, `geometrical_and_steering_doppler`) because I have limited understanding to those.

File: src/compass/utils/lut.py
Comment (LiangJYu) [inline src/compass/utils/lut.py:17]:
```suggestion
    rg_lut: isce.core.LUT2d
```

File: src/compass/utils/lut.py
Comment (LiangJYu) [inline src/compass/utils/lut.py:27]:
```suggestion
    az_lut = burst.bistatic_delay(xstep=xstep, ystep=ystep)
    rg_lut = burst.doppler_induced_range_shiftxstep=xstep, ystep=ystep)
```
More direct; the reassignment of burst corrections computation seem redundant? If the variable names were meant to convey info, I think additonal comments would better suited.
Range correction now a LUT2d instead of numpy array so it's something `isce.geocode.geocode_slc` can immediately use.

File: src/compass/utils/lut.py
Comment (vbrancat) [reply src/compass/utils/lut.py:1]: So, I would suggest that we hold off on that :)

File: src/compass/defaults/s1_cslc_geo.yaml
Comment (vbrancat) [reply src/compass/defaults/s1_cslc_geo.yaml:65]: Well, the primary user would be the `geocodeSlc` core module. However, providing a spacing in seconds in the runconfig/schema seems a bit counter intuitive to me. I am not sure which is the right approach. Shall we go ahead and modify the two corrections in the `s1-reader` to support meters and seconds? Are those two corrections used somewhere else (e.g. RTC)?

File: src/compass/utils/lut.py
Comment (vbrancat) [reply src/compass/utils/lut.py:27]: Well, this is only one element of the look-up tables we will need for the `isce.geocode.geocode_slc`, so the reassignment is not redundant as later on we will have several elements adding up.

Is there a way to go from a numpy.array to a LUT2d using some ISCE3 functionalities?

File: src/compass/defaults/s1_cslc_geo.yaml
Comment (seongsujeong) [inline src/compass/defaults/s1_cslc_geo.yaml:64]:
```suggestion
              # LUT spacing in azimuth direction in seconds
```

File: src/compass/defaults/s1_cslc_radar.yaml
Comment (seongsujeong) [inline src/compass/defaults/s1_cslc_radar.yaml:75]: Is the LUT in radargrid? Then I would suggest to put more reasonable values here. Maybe the same as `s1_cslc_geo.yaml`
```suggestion
           correction_luts:
               # LUT spacing in x/slant range direction in meters
               x_spacing: 120
               # LUT spacing in y/azimuth direction in seconds
               y_spacing: 0.028
```

File: src/compass/schemas/s1_cslc_geo.yaml
Comment (seongsujeong) [inline src/compass/schemas/s1_cslc_geo.yaml:101]:
```suggestion
  # LUT spacing in azimuth direction in seconds
```

File: src/compass/schemas/s1_cslc_geo.yaml
Comment (seongsujeong) [inline src/compass/schemas/s1_cslc_geo.yaml:102]: Not familiar with the schema, but `min=1` means the minimum value(s), or minumum **number** of the value? if the former, I would suggest to decrease it further in accordance with the default value in `s1_cslc_geo.yaml` (i.e. 0.028 seconds)

File: src/compass/schemas/s1_cslc_radar.yaml
Comment (seongsujeong) [inline src/compass/schemas/s1_cslc_radar.yaml:92]: Same as above

**Review by seongsujeong — COMMENTED** [review]
Thanks for lots of discussion about the LUTs. The comments I've just left are nothing but minor issues. See the below for the details.

File: src/compass/schemas/s1_cslc_geo.yaml
Comment (vbrancat) [reply src/compass/schemas/s1_cslc_geo.yaml:102]: Goof catch. I will decrease it to zero so that we can plug in only positive numbers

**Review by seongsujeong — APPROVED** [review]
LGTM. Thanks for addressing the comments I've posted.
Please make sure to resolve the codacy issue before merging.

**Comment by vbrancat** [comment]
@seongsujeong FYI, I cannot resolve the codacy issues at the moment because they are related to the fact that we are not using the `rg_lut` and `az_lut` variables in the code. This will need to be integrated when we release a new version of ISCE3 with conda

**Pull request merged**
